stationsim_model/source/Point2D.cpp

```diff
@@ -20,4 +22,33 @@ namespace station_sim {
     }
 
     Point2D Point2D::operator()(float x, float y) { return Point2D(x, y); }
+
+    // Euclidean distance between two points
+    float Point2D::distance(Point2D p) const {
+        return hypot(this->x - p.x, this->y - p.y);
+    }
+
+    // Euclidean distance between the given and the segment s1-s2.  Based on
+    // https://stackoverflow.com/a/1501725/2442087
+    std::pair<float,Point2D> Point2D::distance_projection(Point2D s1, Point2D s2) const {
+        // Length of segment s1-s2
+        const float l = s1.distance(s2);
+
+        // Fast-track case where s1 == s2
+        if (l == 0.0) {
+            return std::pair<float,Point2D>(this->distance(s1), s1);
+        }
+
+        // Consider the line extending the segment, parameterized as s1 + t (s2 - s1).
```

Comment:
earlier the length of the segment is defined as s1-s2, but here it's s2-s1

Reply:
Yes, the important it to be consistent between the definitions of the parameter `t` and the `projection`.  The length is always non-negative, `s1-s2` or `s2-s1` isn't much relevant here.  In any case your suggestion is to compute `l` as `s2.distance(s1)`?

Reply:
Yes, although like you said since the length is always non-negative this won't change the result from this. Got to say I like the `distance_projection` solution overall!